syntax = "proto3";

import "google/protobuf/empty.proto";
import "enum.proto";
import "message.proto";

package service;

message RunGameRequest {
  string character_1 = 1;
  string character_2 = 2;
  string player_1 = 3;
  string player_2 = 4;
  int32 game_number = 5;
}

message SpectateRequest {}

message SpectatorGameState {
  GrpcFlag state_flag = 1;
  GrpcFrameData frame_data = 2;
  GrpcScreenData screen_data = 3;
  GrpcAudioData audio_data = 4;
  GrpcRoundResult round_result = 5;
}

message InitializeRequest {
  bool player_number = 1;
  string player_name = 2;
  bool is_blind = 3;
}

message InitializeResponse {
  string player_uuid = 1;
}

message ParticipateRequest {
  string player_uuid = 1;
}

message PlayerGameState {
  GrpcFlag state_flag = 1;
  bool is_control = 2;
  GrpcFrameData frame_data = 3;
  GrpcFrameData non_delay_frame_data = 4;
  GrpcScreenData screen_data = 5;
  GrpcAudioData audio_data = 6;
  GrpcGameData game_data = 7;
  GrpcRoundResult round_result = 8;
}

message PlayerInput {
  string player_uuid = 1;
  GrpcKey input_key = 2;
}

service Service {
  rpc RunGame(RunGameRequest) returns (google.protobuf.Empty) {}
  rpc Spectate(SpectateRequest) returns (stream SpectatorGameState) {}

  rpc Initialize(InitializeRequest) returns (InitializeResponse) {}
  rpc Participate(ParticipateRequest) returns (stream PlayerGameState) {}
  rpc Input(PlayerInput) returns (google.protobuf.Empty) {}
}
